hail/python/hailtop/aiocloud/aioterra/azure/fs.py

```diff
@@ -0,0 +1,120 @@
+import asyncio
+import os
+from typing import AsyncContextManager, AsyncIterator, Dict, Set, Tuple, Optional
+
+from hailtop.aiocloud.aioazure import AzureAsyncFS, AzureAsyncFSURL
+from hailtop.aiotools.fs import AsyncFS, ReadableStream, WritableStream, MultiPartCreate, FileStatus, FileListEntry
+from hailtop.utils import time_msecs
+
+
+from .client import TerraClient
+
+
+WORKSPACE_STORAGE_CONTAINER_ID = os.environ.get('WORKSPACE_STORAGE_CONTAINER_ID')
+WORKSPACE_STORAGE_CONTAINER_URL = os.environ.get('WORKSPACE_STORAGE_CONTAINER_URL')
+
+
+class TerraAzureAsyncFS(AsyncFS):
+
+    def __init__(self, **azure_kwargs):
+        self._terra_client = TerraClient()
+        self._azure_fs = AzureAsyncFS(**azure_kwargs)
+        self._sas_token_cache: Dict[str, Tuple[str, int]] = {}
+
+    @property
+    def schemes(self) -> Set[str]:
+        return {'https'}
+
+    @staticmethod
+    def configured() -> bool:
+        return WORKSPACE_STORAGE_CONTAINER_ID is not None and WORKSPACE_STORAGE_CONTAINER_URL is not None
+
+    @staticmethod
+    def valid_url(url: str) -> bool:
+        return AzureAsyncFS.valid_url(url)
+
+    def parse_url(self, url: str) -> AzureAsyncFSURL:
+        return AzureAsyncFS.parse_url(url)
+
+    async def _to_azure_url(self, url: str) -> str:
+        if url in self._sas_token_cache:
+            sas_token, expiration = self._sas_token_cache[url]
+            ten_minutes_from_now = time_msecs() + 10 * 60
+            if expiration > ten_minutes_from_now:
+                return sas_token
+
+        assert WORKSPACE_STORAGE_CONTAINER_URL is not None
+        if url.startswith(WORKSPACE_STORAGE_CONTAINER_URL):
+            sas_token, expiration = await self._create_terra_sas_token(url)
```

Comment:
I think in Scala you update the cache in the creation script; either seems fine but lets be consistent across langs.